EIPS/eip-beacon_state_root_in_evm.md

```diff
@@ -0,0 +1,169 @@
+
+In pseudocode:
+
+```python
+beacon_state_root = block.ommers[0]
+sstore(HISTORY_STORAGE_ADDRESS, block.number, beacon_state_root)
+```
+
+#### New opcode
+
+Beginning at the execution timestamp `FORK_TIMESTAMP`, introduce a new opcode `BEACON_STATE_ROOT` at `OPCODE_VALUE`. This opcode consumes one word from the stack encoding the block number for the root. The opcode has a gas cost of `G_beacon_state_root`.
+
+The result of executing this opcode leaves one word on the stack corresponding to a read of the history contract's storage; in pseudocode:
+
+```python
+block_number = evm.stack.pop()
+sload(HISTORY_STORAGE_ADDRESS, block_number)
+```
+
+If there is no root stored at the requested block number, the opcode follows the existing EVM semantics of `sload` returning `0`.
+
+###  Other considerations
+
+The [Engine API `PayloadAttributes` message](https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payloadattributesv1) will be updated with an additional field to send the relevant beacon state root from the consensus client to the execution client when preparing a block.
+
+This field **MUST** be sent beginning with `FORK_EPOCH` at the consensus layer. Refer to the [consensus specs](https://github.com/ethereum/consensus-specs/) for the definition of `FORK_EPOCH`. (TODO: add full link once defined)
+
+## Rationale
+
+### General strategy
+
+See the rationale for EIP-4399 for discussion about this general strategy of reusing execution block elements for beacon chain data.
+
+### Fork mechanics
+
+This EIP requires the consensus layer and execution layer to execute a network upgrade in lockstep.
+To carry out this task, a `FORK_EPOCH` (of the beacon chain) will be chosen and then used to compute a timestamp `FORK_TIMESTAMP`.
+This `FORK_TIMESTAMP` can be used in the execution layer to identify when the protocol change should be deployed.
+
+This technique works because the timestamps in post-merge execution blocks are aligned to beacon chain slots and thus serve as a proxy for the slot number.
+
+Another option for the fork definition would be to pick a beacon chain epoch and an execution payload block number.
+This design however is not reliable due to the presence of skipped slots on the beacon chain.
+
+### Consensus layer validations
+
+The [`ExecutionPayload` and `ExecutionPayloadHeader`](https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#executionpayload) will be updated with a 32 byte field `beacon_state_root` for the relevant beacon state root for the corresponding execution block.
+This way the beacon state root committed to in the block can be verified against the beacon state root recorded in the beacon chain (e.g. `state.state_roots[state.slot-1]`).
+A future PR to the `consensus-specs` repo with this change is forthcoming.
+
```

Comment:
```suggestion
```
See above, this should be a separate change against the consensus spec and is out of scope of this EIP (which specifies behavior of the execution client).

Reply:
fair  -- this was a hold over from a previous document explaining the "full picture" idea, no problem to drop here now